lib/src/results.dart

```diff
@@ -211,23 +219,21 @@ class _RealmResultsIterator<T extends Object?> implements Iterator<T> {
   int _index;
   T? _current;
 
-  _RealmResultsIterator(RealmResults<T> results)
+  _RealmResultsIterator(RealmResults<T> results, {int index = 0})
       : _results = results,
-        _index = -1;
+        _index = index - 1;
 
   @override
-  T get current => _current as T;
+  T get current => _current ??= _results[_index];
```

Comment:
This will be a problem when we reach the end of the iterator. Previously it would return null, now it will throw.

Reply:
In Dart iteration is controlled by the **bool** returned by `moveNext`. You will never be calling `current` directly. 

If you did you could also make the current implementation throw by calling `current` before `moveNext`.

Notice that no tests where harmed in this refactoring.

Reply:
Also we do actually throw today when `T` is non-nullable.

Reply:
I see, if throwing is consistent with the dart API, then it looks good.

Reply:
Here is a test that illustrates the behaviour:
```dart
import 'package:test/test.dart';

void main(List<String> arguments) {
  test('iterable', () {
    final list = [1];
    final it = list.iterator;
    // you are not supposed to call current before first moveNext
    expect(() => it.current, throwsA(isA<TypeError>()));
    expect(it.moveNext(), isTrue);
    expect(it.moveNext(), isFalse);
    // you are not supposed to call current, if moveNext return false
    expect(() => it.current, throwsA(isA<TypeError>()));
  });
}
```

Reply:
I have added a test for these edge-cases for `_RealmResultsIterator` as well